uefi/src/table/boot.rs

```diff
@@ -2082,6 +2079,104 @@ pub struct MemoryMapSize {
     pub map_size: usize,
 }
 
+/// An iterator of [`MemoryDescriptor`] which returns elements in sorted order. The underlying memory map is always
+/// associated with the unique [`MemoryMapKey`] contained in the struct.
+pub struct MemoryMap<'buf> {
+    key: MemoryMapKey,
+    buf: &'buf mut [u8],
+    entry_size: usize,
+    len: usize,
+}
+
+impl<'buf> MemoryMap<'buf> {
+    #[allow(clippy::must_use_candidate)]
+    /// Returns the unique [`MemoryMapKey`] associated with the memory map
+    pub fn key(&self) -> MemoryMapKey {
+        self.key
+    }
+
+    /// Sorts the memory map in place
+    pub fn sort(&mut self) {
+        unsafe {
+            self.qsort(0, self.len - 1);
```

Comment:
Why can't we use some built-in basic sorting functionality? It seems overkill to implement quick sort, doesn't it? What do you think, @nicholasbishop ?

I thought about a basic bubble sort as a UEFI memory map is unlikely to have thousands of entries.

Reply:
I don't think the built-in sorting functions allow an unspecified constant spacing between elements in a slice, which would complicate using any of the built in basic sorting functions

Reply:
What do you think about pulling in a dep on the [index-sort](https://docs.rs/index-sort) crate? It's a small crate with a very short commit history, but it seems like exactly what we need for this use case. I would prefer that to implementing any of our own sorts, as that seems somewhat outside the scope of a uefi crate.

(I was a little surprised that I couldn't find more crates related to sorting non-slice data structures, but I guess it is kind of a niche need.)

Reply:
Oh oops, just realized that crate isn't no-std. I might put up a PR for that.

Reply:
OK, I put up a PR in that project to make it no-std compatible.

I don't want to block this PR on it though. I do feel a little uncomfortable about adding a somewhat complicated sort implementation here, but I'm OK with it for now because we can hopefully replace it in the future with a dependency on an external crate.

Reply:
For the record, this is the PR nicholas set up: https://github.com/arlencox/index-sort/pull/1